pkg/virt-handler/node-labeller/node_labeller.go

```diff
@@ -0,0 +1,208 @@
+	"time"
+
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+
+	"kubevirt.io/client-go/kubecli"
+	"kubevirt.io/client-go/log"
+	device_manager "kubevirt.io/kubevirt/pkg/virt-handler/device-manager"
+)
+
+const (
+	labelNamespace    = "feature.node.kubernetes.io"
+	labellerNamespace = "node-labeller"
+)
+
+var nodeLabellerVolumePath = "/var/lib/kubevirt-node-labeller"
+
+//NodeLabeller struct holds informations needed to run node-labeller
+type NodeLabeller struct {
+	kvmController     *device_manager.DeviceController
+	configMapInformer cache.SharedIndexInformer
+	clientset         kubecli.KubevirtClient
+	Queue             workqueue.RateLimitingInterface
+	host              string
+	namespace         string
+}
+
+func NewNodeLabeller(kvmController *device_manager.DeviceController, configMapInformer cache.SharedIndexInformer, clientset kubecli.KubevirtClient, host, namespace string) *NodeLabeller {
+	return &NodeLabeller{
+		kvmController:     kvmController,
+		configMapInformer: configMapInformer,
+		clientset:         clientset,
+		Queue:             workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter()),
+
+		host:      host,
+		namespace: namespace,
+	}
+}
+
+//Run runs node-labeller
+func (n *NodeLabeller) Run(threadiness int, stop chan struct{}) {
+	defer n.Queue.ShutDown()
+	logger := log.DefaultLogger()
+	logger.Infof("node-labeller is running")
+
+	n.configMapInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			n.Queue.AddRateLimited(obj)
+		},
+		DeleteFunc: func(obj interface{}) {
+			n.Queue.AddRateLimited(obj)
+		},
+		UpdateFunc: func(old, new interface{}) {
+			n.Queue.AddRateLimited(new)
+		},
+	})
+
+	go n.configMapInformer.Run(stop)
+	cache.WaitForCacheSync(stop, n.configMapInformer.HasSynced)
+
+	for i := 0; i < threadiness; i++ {
+		go wait.Until(n.runWorker, time.Second, stop)
+	}
+	<-stop
+}
+func (n *NodeLabeller) runWorker() {
+	for n.Execute() {
+	}
+}
+
+func (n *NodeLabeller) Execute() bool {
+	key, quit := n.Queue.Get()
+	if quit {
+		return false
+	}
+	defer n.Queue.Done(key)
+	err := n.run()
+
+	if err != nil {
+		n.Queue.AddRateLimited(key)
+	} else {
+		n.Queue.Forget(key)
+	}
+	return true
+}
+
+func (n *NodeLabeller) run() error {
+	logger := log.DefaultLogger()
+
+	if !n.kvmController.NodeHasDevice(device_manager.KVMPath) {
+		return goerror.New(fmt.Sprintf("Node-labeller cannot work without KVM device."))
+	}
+
+	cpuFeatures := make(map[string]bool)
+	cpuModels := make([]string, 0)
+	//parse all informations
+	cpuModels, cpuFeatures, err := n.getCPUInfo()
+	if err != nil {
+		logger.Infof("node-labeller cannot get new labels %s\n", err.Error())
+		return err
+	}
+	//get node on which virt-handler is running
+	node, err := n.clientset.CoreV1().Nodes().Get(n.host, metav1.GetOptions{})
```

Comment:
we should be getting the node from cache instead of performing the Get here. the controller needs a node informer for us to do that.